File: app/test/sync/crdt_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:gullak/sync/crdt.dart';

void main() {
  final vectors =
      jsonDecode(File('../sync_test_vectors/crdt_v1.json').readAsStringSync())
          as Map<String, dynamic>;

  group('shared canonical vectors', () {
    test('vector format is the frozen v1 corpus', () {
      expect(vectors['format'], 'gullak-crdt-v1');
    });

    for (final raw in vectors['registerCases'] as List<dynamic>) {
      final vector = raw as Map<String, dynamic>;
      test(vector['name'] as String, () {
        final state = _vectorState(
          vector['changes'] as List<dynamic>,
          target: vector['target'] as Map<String, dynamic>,
        );
        expect(state.canonicalJson(), vector['expectedStateJson']);
        expect(
          encodeCanonicalJson(state.visibleWinner?.value),
          encodeCanonicalJson(vector['expectedValue']),
        );
        expect(state.conflictCount, vector['expectedConflictCount']);
      });
    }

    for (final raw in vectors['lifecycleCases'] as List<dynamic>) {
      final vector = raw as Map<String, dynamic>;
      test('lifecycle: ${vector['name']}', () {
        expect(
          removeWinsExistence(_vectorState(vector['changes'] as List<dynamic>)),
          vector['expected'],
        );
      });
    }

    for (final raw in vectors['membershipCases'] as List<dynamic>) {
      final vector = raw as Map<String, dynamic>;
      test('membership: ${vector['name']}', () {
        expect(
          addWinsMembership(_vectorState(vector['changes'] as List<dynamic>)),
          vector['expected'],
        );
      });
    }
  });

  group('causal MVR algebra', () {
    test('merge is commutative, associative and idempotent', () {
      final a = _candidate(actor: 'a', sequence: 1, lamport: 1, value: 'A');
      final b = _candidate(actor: 'b', sequence: 1, lamport: 2, value: 'B');
      final c = _candidate(actor: 'c', sequence: 1, lamport: 3, value: 'C');
      final left = RegisterState([a]);
      final right = RegisterState([b]);
      final third = RegisterState([c]);

      expect(
        left.merge(right).canonicalJson(),
        right.merge(left).canonicalJson(),
      );
      expect(
        left.merge(right).merge(third).canonicalJson(),
        left.merge(right.merge(third)).canonicalJson(),
      );
      expect(left.merge(left).canonicalJson(), left.canonicalJson());
    });

    test('all permutations and duplicates converge', () {
      final candidates = [
        _candidate(actor: 'a', sequence: 1, lamport: 1, value: 'A'),
        _candidate(actor: 'b', sequence: 1, lamport: 4, value: 'B'),
        _candidate(
          actor: 'a',
          sequence: 2,
          context: const {'a': 1, 'b': 1},
          lamport: 5,
          value: 'resolved',
        ),
      ];
      final expected = RegisterState(candidates).canonicalJson();
      for (final permutation in _permutations(candidates)) {
        var state = RegisterState();
        for (final candidate in [...permutation, ...permutation]) {
          state = state.add(candidate);
        }
        expect(state.canonicalJson(), expected);
      }
    });

    test('causal successor removes every observed candidate', () {
      final a = _candidate(actor: 'a', sequence: 1, lamport: 10, value: 'A');
      final b = _candidate(actor: 'b', sequence: 1, lamport: 20, value: 'B');
      final resolution = _candidate(
        actor: 'a',
        sequence: 2,
        context: const {'a': 1, 'b': 1},
        lamport: 21,
        value: 'resolved',
      );

      final state = RegisterState([resolution, b, a]);
      expect(state.candidates, hasLength(1));
      expect(state.visibleWinner?.value, 'resolved');
      expect(state.hasConflict, isFalse);
    });

    test('concurrent values remain while winner is deterministic', () {
      final a = _candidate(
        actor: 'a',
        sequence: 7,
        context: const {'a': 6},
        lamport: 11,
        value: 'A',
      );
      final b = _candidate(
        actor: 'b',
        sequence: 2,
        context: const {'b': 1},
        lamport: 11,
        value: 'B',
      );
      final state = RegisterState([b, a]);

      expect(state.hasConflict, isTrue);
      expect(state.conflict.candidates, hasLength(2));
      expect(state.visibleWinner?.value, 'B');
      expect(state.conflict.winner?.actorId, 'b');
    });

    test('wall clock is ignored by causal merge and visible winner', () {
      final future = _candidate(
        actor: 'a',
        sequence: 1,
        lamport: 1,
        wallTimeMs: 9000000000000,
        value: 'future clock',
      );
      final past = _candidate(
        actor: 'b',
        sequence: 1,
        lamport: 2,
        wallTimeMs: 0,
        value: 'past clock',
      );

      expect(RegisterState([future, past]).visibleWinner?.value, 'past clock');
    });
  });

  group('lifecycle and membership policies', () {
    test('remove wins concurrent lifecycle assignments', () {
      final add = _candidate(field: r'$exists', actor: 'a', value: true);
      final remove = _candidate(field: r'$exists', actor: 'b', value: false);
      expect(removeWinsExistence(RegisterState([add, remove])), isFalse);
    });

    test('causal restore supersedes remove', () {
      final remove = _candidate(field: r'$exists', actor: 'a', value: false);
      final restore = _candidate(
        field: r'$exists',
        actor: 'b',
        context: const {'a': 1},
        lamport: 2,
        value: true,
      );
      expect(removeWinsExistence(RegisterState([restore, remove])), isTrue);
    });

    test('add wins concurrent membership but observed remove wins later', () {
      final add = _candidate(field: r'$exists', actor: 'a', value: true);
      final concurrentRemove = _candidate(
        field: r'$exists',
        actor: 'b',
        value: false,
      );
      expect(addWinsMembership(RegisterState([add, concurrentRemove])), isTrue);

      final observedRemove = _candidate(
        field: r'$exists',
        actor: 'b',
        sequence: 2,
        context: const {'a': 1, 'b': 1},
        lamport: 2,
        value: false,
      );
      expect(
        addWinsMembership(
          RegisterState([add, concurrentRemove, observedRemove]),
        ),
        isFalse,
      );
    });
  });

  group('wire validation and canonicalization', () {
    test('round trips null and unknown fields opaquely', () {
      final envelope = ChangeEnvelope(
        epoch: 'epoch',
        changeId: 'phone:1',
        actorId: 'phone',
        sequence: 1,
        context: const {},
        lamport: 1,
        wallTimeMs: 123,
        schemaVersion: 99,
        ops: [
          AssignOp(
            resource: 'transactions',
            entityId: 'txn',
            field: 'future.metadata',
            value: const {
              'nullable': null,
              'nested': [1, true, 'x'],
            },
          ),
        ],
      );

      final decoded = ChangeEnvelope.fromJson(
        Map<String, Object?>.from(jsonDecode(envelope.canonicalJson()) as Map),
      );
      expect(decoded.canonicalJson(), envelope.canonicalJson());
      expect(decoded.candidates.single.value, envelope.ops.single.value);
    });

    test('canonical register JSON has stable candidate and map ordering', () {
      final a = _candidate(actor: 'z', value: const {'z': 1, 'a': 2});
      final b = _candidate(actor: 'a', value: null);
      final canonical = RegisterState([a, b]).canonicalJson();
      expect(canonical, RegisterState([b, a]).canonicalJson());

      final decoded = RegisterState.fromJson(
        Map<String, Object?>.from(jsonDecode(canonical) as Map),
        resource: 'transactions',
        entityId: 'txn',
        field: 'notes',
      );
      expect(decoded.canonicalJson(), canonical);
      expect(decoded.visibleWinner?.value, const {'z': 1, 'a': 2});
    });

    test('rejects malformed envelopes and dot reuse with different bytes', () {
      expect(
        () => ChangeEnvelope(
          epoch: 'epoch',
          changeId: 'wrong',
          actorId: 'phone',
          sequence: 1,
          context: const {},
          lamport: 1,
          wallTimeMs: 0,
          schemaVersion: 1,
          ops: [AssignOp(resource: 'r', entityId: 'e', field: 'f', value: 1)],
        ),
        throwsA(isA<CrdtValidationException>()),
      );

      final first = _candidate(actor: 'a', value: 1);
      final reused = _candidate(actor: 'a', value: 2);
      expect(
        () => RegisterState([first, reused]),
        throwsA(isA<CrdtValidationException>()),
      );
    });

    test('rejects zero/gapped self contexts and causal cycles', () {
      expect(
        () => _candidate(actor: 'a', sequence: 2, context: const {'a': 0}),
        throwsA(isA<CrdtValidationException>()),
      );
      expect(
        () => _candidate(actor: 'a', sequence: 3, context: const {'a': 1}),
        throwsA(isA<CrdtValidationException>()),
      );

      final a = _candidate(actor: 'a', context: const {'b': 1});
      final b = _candidate(actor: 'b', context: const {'a': 1});
      expect(
        () => RegisterState([a, b]),
        throwsA(isA<CrdtValidationException>()),
      );
    });

    test('rejects causal cycles of three or more candidates', () {
      final a = _candidate(actor: 'a', context: const {'b': 1});
      final b = _candidate(actor: 'b', context: const {'c': 1});
      final c = _candidate(actor: 'c', context: const {'a': 1});
      expect(
        () => RegisterState([a, b, c]),
        throwsA(isA<CrdtValidationException>()),
      );
    });

    test('numeric JSON is safe and cross-runtime canonical', () {
      expect(
        () => AssignOp(
          resource: 'r',
          entityId: 'e',
          field: 'f',
          value: 9007199254740992,
        ),
        throwsA(isA<CrdtValidationException>()),
      );
      expect(
        () => encodeCanonicalJson(181.25),
        throwsA(isA<CrdtValidationException>()),
      );
      expect(
        () => encodeCanonicalJson(12.971598765),
        throwsA(isA<CrdtValidationException>()),
      );
      expect(
        encodeCanonicalJson({'lat': 12.9715987, 'lng': -77.5945623}),
        '{"lat":12.9715987,"lng":-77.5945623}',
      );
      expect(encodeCanonicalJson(-0.0), '0');
    });
  });
}

RegisterState _vectorState(
  List<dynamic> changes, {
  Map<String, dynamic>? target,
}) {
  var state = RegisterState();
  for (final raw in changes) {
    final envelope = ChangeEnvelope.fromJson(
      Map<String, Object?>.from(raw as Map),
    );
    final op = target == null
        ? envelope.ops.first
        : envelope.ops.singleWhere(
            (candidate) =>
                candidate.resource == target['resource'] &&
                candidate.entityId == target['entityId'] &&
                candidate.field == target['field'],
          );
    final candidate = envelope.candidates.singleWhere(
      (candidate) =>
          candidate.resource == op.resource &&
          candidate.entityId == op.entityId &&
          candidate.field == op.field,
    );
    state = state.add(candidate);
  }
  return state;
}

Candidate _candidate({
  String resource = 'transactions',
  String entityId = 'txn',
  String field = 'notes',
  required String actor,
  int sequence = 1,
  Map<String, int> context = const {},
  int lamport = 1,
  int wallTimeMs = 0,
  Object? value,
}) => Candidate(
  changeId: '$actor:$sequence',
  actorId: actor,
  sequence: sequence,
  context: context,
  lamport: lamport,
  wallTimeMs: wallTimeMs,
  schemaVersion: 1,
  resource: resource,
  entityId: entityId,
  field: field,
  value: value,
);

Iterable<List<T>> _permutations<T>(List<T> values) sync* {
  if (values.length < 2) {
    yield List<T>.from(values);
    return;
  }
  for (var index = 0; index < values.length; index++) {
    final head = values[index];
    final tail = List<T>.from(values)..removeAt(index);
    for (final permutation in _permutations(tail)) {
      yield [head, ...permutation];
    }
  }
}

File: pi-server/src/sync/crdt.test.ts
import { readFileSync } from "node:fs";

import { describe, expect, test } from "vitest";

import {
  type AssignOp,
  type Candidate,
  type ChangeEnvelope,
  type JsonValue,
  type RegisterState,
  CrdtValidationError,
  candidateFor,
  canonicalChangeJson,
  canonicalRegisterJson,
  compareCandidates,
  foldRegisterCandidates,
  mergeCandidate,
  mergeRegisterStates,
  normalizeRegisterState,
  projectAddWinsMembership,
  projectRegister,
  projectRemoveWinsExists,
  stableJson,
  validateChangeEnvelope,
} from "./crdt.ts";

type Target = Pick<AssignOp, "resource" | "entityId" | "field">;

type RegisterVector = {
  name: string;
  target: Target;
  changes: unknown[];
  expectedStateJson: string;
  expectedValue: JsonValue;
  expectedConflictCount: number;
};

type BooleanVector = {
  name: string;
  changes: unknown[];
  expected: boolean;
};

type Vectors = {
  format: string;
  registerCases: RegisterVector[];
  lifecycleCases: BooleanVector[];
  membershipCases: BooleanVector[];
};

const vectors = JSON.parse(
  readFileSync(
    new URL("../../../sync_test_vectors/crdt_v1.json", import.meta.url),
    "utf8",
  ),
) as Vectors;

function matchingCandidate(changeValue: unknown, target?: Target): Candidate {
  const change = validateChangeEnvelope(changeValue);
  const op =
    target === undefined
      ? change.ops[0]
      : change.ops.find(
          (item) =>
            item.resource === target.resource &&
            item.entityId === target.entityId &&
            item.field === target.field,
        );
  if (op === undefined) throw new Error("test vector has no matching op");
  return candidateFor(change, op);
}

function vectorState(changes: unknown[], target?: Target): RegisterState {
  return foldRegisterCandidates(
    changes.map((change) => matchingCandidate(change, target)),
  );
}

function permutations<T>(items: T[]): T[][] {
  if (items.length < 2) return [items];
  return items.flatMap((item, index) =>
    permutations([...items.slice(0, index), ...items.slice(index + 1)]).map(
      (rest) => [item, ...rest],
    ),
  );
}

describe("shared canonical vectors", () => {
  test("vector format is the frozen v1 corpus", () => {
    expect(vectors.format).toBe("gullak-crdt-v1");
  });

  for (const vector of vectors.registerCases) {
    test(vector.name, () => {
      const state = vectorState(vector.changes, vector.target);
      const projection = projectRegister(state);

      expect(canonicalRegisterJson(state)).toBe(vector.expectedStateJson);
      expect(stableJson(projection.value)).toBe(
        stableJson(vector.expectedValue),
      );
      expect(projection.conflict?.alternatives.length ?? 0).toBe(
        vector.expectedConflictCount,
      );
    });
  }

  for (const vector of vectors.lifecycleCases) {
    test(`lifecycle: ${vector.name}`, () => {
      expect(projectRemoveWinsExists(vectorState(vector.changes))).toBe(
        vector.expected,
      );
    });
  }

  for (const vector of vectors.membershipCases) {
    test(`membership: ${vector.name}`, () => {
      expect(projectAddWinsMembership(vectorState(vector.changes))).toBe(
        vector.expected,
      );
    });
  }
});

describe("register merge laws", () => {
  const a: Candidate = {
    dot: { actorId: "a", sequence: 1 },
    context: {},
    lamport: 1,
    value: "a1",
  };
  const b: Candidate = {
    dot: { actorId: "b", sequence: 1 },
    context: {},
    lamport: 2,
    value: "b1",
  };
  const c: Candidate = {
    dot: { actorId: "c", sequence: 1 },
    context: { a: 1, b: 1 },
    lamport: 3,
    value: "c1",
  };
  const state = (candidate: Candidate): RegisterState => ({
    candidates: [candidate],
  });

  test("merge is commutative", () => {
    expect(canonicalRegisterJson(mergeRegisterStates(state(a), state(b)))).toBe(
      canonicalRegisterJson(mergeRegisterStates(state(b), state(a))),
    );
  });

  test("merge is associative", () => {
    const left = mergeRegisterStates(
      mergeRegisterStates(state(a), state(b)),
      state(c),
    );
    const right = mergeRegisterStates(
      state(a),
      mergeRegisterStates(state(b), state(c)),
    );
    expect(canonicalRegisterJson(left)).toBe(canonicalRegisterJson(right));
  });

  test("merge is idempotent", () => {
    const once = mergeRegisterStates(state(a), state(b));
    expect(canonicalRegisterJson(mergeRegisterStates(once, once))).toBe(
      canonicalRegisterJson(once),
    );
  });

  test("every permutation and duplicate delivery has identical state", () => {
    const expected = canonicalRegisterJson(foldRegisterCandidates([a, b, c]));
    for (const order of permutations([a, b, c])) {
      expect(canonicalRegisterJson(foldRegisterCandidates(order))).toBe(
        expected,
      );
      expect(
        canonicalRegisterJson(
          foldRegisterCandidates([order[0] as Candidate, ...order]),
        ),
      ).toBe(expected);
    }
  });
});

describe("causal and projection semantics", () => {
  test("a causal successor removes the predecessor from the antichain", () => {
    const old: Candidate = {
      dot: { actorId: "phone", sequence: 1 },
      context: {},
      lamport: 100,
      value: "old",
    };
    const next: Candidate = {
      dot: { actorId: "server", sequence: 1 },
      context: { phone: 1 },
      lamport: 101,
      value: "new",
    };
    const state = foldRegisterCandidates([next, old]);
    expect(state.candidates).toHaveLength(1);
    expect(projectRegister(state).value).toBe("new");
  });

  test("concurrent same-field assignments are retained and reported", () => {
    const vector = vectors.registerCases.find((item) =>
      item.name.startsWith("concurrent values"),
    );
    if (vector === undefined) throw new Error("missing conflict vector");
    const projection = projectRegister(
      vectorState(vector.changes, vector.target),
    );
    expect(projection.winner?.dot.actorId).toBe("web");
    expect(projection.conflict?.alternatives).toHaveLength(1);
    expect(projection.conflict?.alternatives[0]?.value).toBe("phone note");
  });

  test("wall time has no effect on a candidate or visible winner", () => {
    const vector = vectors.registerCases.find((item) =>
      item.name.startsWith("concurrent values"),
    );
    if (vector === undefined) throw new Error("missing clock vector");
    const changes = vector.changes.map((item) => structuredClone(item));
    const first = changes[0] as Record<string, unknown>;
    const second = changes[1] as Record<string, unknown>;
    first.wallTimeMs = 0;
    second.wallTimeMs = 8_000_000_000_000;

    const before = vectorState(vector.changes, vector.target);
    const after = vectorState(changes, vector.target);
    expect(canonicalRegisterJson(after)).toBe(canonicalRegisterJson(before));
    expect(projectRegister(after).value).toBe("web note");
  });

  test("winner tuple is lamport, actorId, then sequence", () => {
    const base: Candidate = {
      dot: { actorId: "a", sequence: 9 },
      context: { a: 8 },
      lamport: 3,
      value: null,
    };
    expect(
      compareCandidates(base, {
        ...base,
        dot: { actorId: "z", sequence: 1 },
        context: {},
      }),
    ).toBeLessThan(0);
    expect(
      compareCandidates(base, {
        ...base,
        dot: { actorId: "a", sequence: 10 },
        context: { a: 9 },
      }),
    ).toBeLessThan(0);
    expect(compareCandidates(base, { ...base, lamport: 4 })).toBeLessThan(0);
  });

  test("non-boolean lifecycle and membership candidates fail closed", () => {
    const state: RegisterState = {
      candidates: [
        {
          dot: { actorId: "phone", sequence: 1 },
          context: {},
          lamport: 1,
          value: "yes",
        },
      ],
    };
    expect(() => projectRemoveWinsExists(state)).toThrow(CrdtValidationError);
    expect(() => projectAddWinsMembership(state)).toThrow(CrdtValidationError);
  });
});

describe("validation and canonical encoding", () => {
  const valid = (): ChangeEnvelope => ({
    protocol: 2,
    epoch: "epoch",
    changeId: "phone:2",
    actorId: "phone",
    sequence: 2,
    context: { phone: 1 },
    lamport: 2,
    wallTimeMs: 1,
    schemaVersion: 1,
    ops: [
      {
        kind: "assign",
        resource: "transactions",
        entityId: "t1",
        field: "notes",
        value: null,
      },
    ],
  });

  test("dot identity, self context, and JSON values are validated", () => {
    expect(() =>
      validateChangeEnvelope({ ...valid(), changeId: "phone:9" }),
    ).toThrow(/actorId:sequence/);
    expect(() =>
      validateChangeEnvelope({ ...valid(), context: { phone: 2 } }),
    ).toThrow(/sequence - 1/);
    expect(() =>
      validateChangeEnvelope({ ...valid(), context: { phone: 1, web: 0 } }),
    ).toThrow(/safe integer/);
    expect(() =>
      validateChangeEnvelope({
        ...valid(),
        ops: [{ ...valid().ops[0], value: Number.NaN }],
      }),
    ).toThrow(/finite JSON/);
  });

  test("one atomic change cannot assign one register twice", () => {
    expect(() =>
      validateChangeEnvelope({
        ...valid(),
        ops: [valid().ops[0], valid().ops[0]],
      }),
    ).toThrow(/same register/);
  });

  test("same dot with different bytes is rejected", () => {
    const first = matchingCandidate(valid());
    expect(() =>
      mergeRegisterStates(
        { candidates: [first] },
        { candidates: [{ ...first, value: "different" }] },
      ),
    ).toThrow(/same dot/);
  });

  test("causal cycles are rejected", () => {
    const a: Candidate = {
      dot: { actorId: "a", sequence: 1 },
      context: { b: 1 },
      lamport: 1,
      value: "a",
    };
    const b: Candidate = {
      dot: { actorId: "b", sequence: 1 },
      context: { a: 1 },
      lamport: 1,
      value: "b",
    };
    expect(() => mergeCandidate({ candidates: [a] }, b)).toThrow(
      /causal cycle/,
    );
  });

  test("causal cycles of three or more candidates are rejected", () => {
    const a: Candidate = {
      dot: { actorId: "a", sequence: 1 },
      context: { b: 1 },
      lamport: 1,
      value: "a",
    };
    const b: Candidate = {
      dot: { actorId: "b", sequence: 1 },
      context: { c: 1 },
      lamport: 1,
      value: "b",
    };
    const c: Candidate = {
      dot: { actorId: "c", sequence: 1 },
      context: { a: 1 },
      lamport: 1,
      value: "c",
    };
    expect(() => normalizeRegisterState({ candidates: [a, b, c] })).toThrow(
      /causal cycle/,
    );
  });

  test("numeric JSON is safe and cross-runtime canonical", () => {
    expect(() => stableJson(9_007_199_254_740_992)).toThrow(/safe integers/);
    expect(() => stableJson(181.25)).toThrow(/portable decimals/);
    expect(() => stableJson(12.971598765)).toThrow(/portable decimals/);
    expect(stableJson({ lat: 12.9715987, lng: -77.5945623 })).toBe(
      '{"lat":12.9715987,"lng":-77.5945623}',
    );
    expect(stableJson(-0)).toBe("0");
  });

  test("prototype-shaped actor ids remain ordinary context entries", () => {
    const envelope = validateChangeEnvelope(
      JSON.parse(`{
        "protocol":2,
        "epoch":"epoch",
        "changeId":"__proto__:2",
        "actorId":"__proto__",
        "sequence":2,
        "context":{"__proto__":1},
        "lamport":2,
        "wallTimeMs":1,
        "schemaVersion":1,
        "ops":[{"kind":"assign","resource":"transactions","entityId":"t","field":"notes","value":"safe"}]
      }`),
    );
    expect(
      Object.prototype.hasOwnProperty.call(envelope.context, "__proto__"),
    ).toBe(true);
    expect(envelope.context.__proto__).toBe(1);
    expect(canonicalChangeJson(envelope)).toContain('"__proto__":1');
  });

  test("canonical JSON recursively sorts keys and canonicalizes changes", () => {
    expect(stableJson({ z: null, a: { y: 2, x: 1 } })).toBe(
      '{"a":{"x":1,"y":2},"z":null}',
    );
    const shuffled = {
      ops: valid().ops,
      wallTimeMs: 1,
      lamport: 2,
      context: { phone: 1 },
      sequence: 2,
      actorId: "phone",
      changeId: "phone:2",
      schemaVersion: 1,
      epoch: "epoch",
      protocol: 2,
    };
    expect(canonicalChangeJson(shuffled)).toBe(canonicalChangeJson(valid()));
  });
});
